File: algorithm_introduction/ss4_divide_and_conquer/python/square_matrix_multiply.py
def square_matrix_multiply(A, B):
    n = len(A)
    C = [[0 for _ in range(n)] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            for k in range(n):
                C[i][j] += A[i][k] * B[k][j]
    return C


def main():
    A = [
        [0, 1],
        [2, 3]
    ]
    B = [
        [4, 5],
        [6, 7]
    ]

    res = square_matrix_multiply(A, B)
    print(res)


if __name__ == '__main__':
    main()

File: algorithm_introduction/ss04_divide_and_conquer/cpp/square_matrix_multiply.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> square_matrix_multiply(int A[][2], int B[][2]) {
  int n = sizeof(A[0]) / sizeof(A[0][0]);
  vector<vector<int>> res(n, vector<int>(n, 0));
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < n; ++j) {
      for (int k = 0; k < n; ++k) {
        res[i][j] += A[i][k] * B[k][j];
      }
    }
  }
  return res;
}

int main() {
  int A[2][2] = {{0, 1}, {2, 3}};
  int B[2][2] = {{4, 5}, {6, 7}};
  vector<vector<int>> res(2, vector<int>(2, 0));

  res = square_matrix_multiply(A, B);
  cout << "[\n";
  for (int i = 0; i < 2; ++i) {
    cout << "  [";
    for (int j = 0; j < 2; ++j) {
      cout << res[i][j] << ", ";
    }
    cout << "]\n";
  }
  cout << "]" << endl;
}